haystack/components/builders/hypothetical_doc_query.py

```diff
@@ -0,0 +1,127 @@
+from haystack import Pipeline, component, Document
+from haystack.components.embedders.sentence_transformers_document_embedder import SentenceTransformersDocumentEmbedder
+from haystack.components.generators.openai import OpenAIGenerator
+from haystack.components.builders import PromptBuilder
+
+from typing import List, Dict, Any
+from numpy import array, mean
+
+
+@component
+class InstructLM:
+    """
+    Subcomponent to generate hypothetical documents using an instruction-following language model (InstructLM).
+    """
+
+    def __init__(self, instruct_llm, nr_completions):
+        self.prompt_builder = PromptBuilder(
+            template="""
+        Given a question, generate a paragraph of text that answers the question.
+        Question: {{question}}
+        Paragraph:
+        """
+        )
+        self.generator = OpenAIGenerator(
+            model=instruct_llm, generation_kwargs={"n": nr_completions, "temperature": 0.75, "max_tokens": 400}
+        )
+
+        self.pipeline = Pipeline()
```

Comment:
Let's discuss the topic of supercomponents/subpipelines separately. The question here is how many of the parameters of these inner components we want to expose. For example, `api_key` we should expose too.

Reply:
I wrote something about on the proposal/discussion - see the link below